src/cmd-cloud-prune

```diff
@@ -0,0 +1,303 @@
+
+Build = collections.namedtuple("Build", ["id", "timestamp", "images", "arch"])
+# set metadata caching to 5m
+CACHE_MAX_AGE_METADATA = 60 * 5
+
+
+def main():
+    args = parse_args()
+    bucket, prefix = args.url.split("/", 1)
+    stream = args.stream
+    # Boto3 loads credentials from ~/.aws/config by default and we can change
+    # this default location by setting the AWS_CONFIG_FILE environment variable.
+    # The alternative is to manually pass ACCESS_KEY and SECRET_KEY which isn't favourable.
+    if args.aws_config_file:
+        os.environ["AWS_CONFIG_FILE"] = args.aws_config_file
+
+    gcp_cloud_config = {
+        "gcp": {
+            "json-key": args.gcp_json_key,
+            "project": args.gcp_project,
+        }
+    }
+    # These lists are up to date as of schema hash
+    # 4c19aed3b3d84af278780bff63728510bb3e70613e4c4eef8cabd7939eb31bd8. If changing
+    # this hash, ensure that the list of supported and unsupported artifacts below
+    # is up to date.
+    supported = ["amis", "gcp"]
+    unsupported = []
+
+    with open(f"{args.policy}") as f:
+        policy = yaml.safe_load(f)
+
+    if stream not in policy:
+        print(f"No pruning policy specified for {stream} stream in policy.yaml")
+        return
+    else:
+        s3_client = boto3.client("s3")
+        # If the build key is set in the policy file, then the cloud-uploads key must
+        # also be present, and the duration of cloud-uploads must be equal or shorter
+        if "build" in policy[stream].keys():
+            cloud_uploads_check(policy[stream])
+
+        # Base URL for Fedora CoreOS builds
+
+        buildJsonData = get_json_from_s3(bucket, prefix + "/builds/builds.json")
+        if buildJsonData is None:
+            print(f"No builds.json found for {stream} in {bucket} bucket")
+            return
+
+        # action is basically whatever is needed to be pruned for the respective stream
+        for action in policy[stream]:
+            duration = policy[stream][action]
+            duration_in_months = get_months(duration)
+            refDate = datetime.datetime.now() - relativedelta(months=int(duration_in_months))
+
+            print(f"Pruning {duration} old {action} for {stream} builds")
+            # Enumerating in reverse to go from the oldest build to the newest one
+            for index, build in enumerate(reversed(buildJsonData["builds"])):
+                build_id = build["id"]
+                if "policy-cleanup" in build.keys():
+                    # If we have already pruned the spedified resources for this
+                    # build as per builds.json, we skip through it.
+                    if action in build["policy-cleanup"]:
+                        print(f"The {build_id} has already had {action} pruning completed")
+                        continue
+                timestamp = build_id.split('.')[1]
+                buildDate = datetime.datetime(int(timestamp[0:4]), int(timestamp[4:6]), int(timestamp[-2:]))
```

Comment:
Not super happy about this since it's of course FCOS-specific, and it doesn't feel great to auto-detect either. I think in the future we could have a new e.g. `--timestamp-source` that takes e.g. `fcos-version`, `rhcos-version`, or `build-meta`. The latter of which would actually fetch the `meta.json`.

Or... we enhance the `builds.json` schema to also have the build timestamp for each build. Really tempting...

Anyway, for now OK to assume FCOS versioning. I think let's factor this out into a function called e.g. `parse_fcos_version()`. It should parse the whole version string or bail if it can't. Use https://github.com/coreos/fedora-coreos-config/blob/testing-devel/versionary#L156 as a starting function definition and adapt that.